include/sparrow/optional.hpp

```diff
@@ -0,0 +1,500 @@
+        requires (
+            util::both_constructible_from_cond_ref<T, TO, B, BO> and
+            not util::initializable_from_refs<T, optional<TO, BO>>
+        )
+        explicit(not util::both_convertible_from_cond_ref<T, TO, B, BO>)
+        constexpr optional(optional<TO, BO>&& rhs)
+            : m_value(std::move(rhs.m_value))
+            , m_flag(std::move(rhs.m_flag))
+        {
+        }
+
+        constexpr optional(value_type&& value, flag_type&& flag)
+            : m_value(std::move(value))
+            , m_flag(std::move(flag))
+        {
+        }
+
+        constexpr optional(std::add_lvalue_reference_t<T> value, std::add_lvalue_reference_t<B> flag)
+            : m_value(value)
+            , m_flag(flag)
+        {
+        }
+
+        constexpr optional(value_type&& value, std::add_lvalue_reference_t<B> flag)
+            : m_value(std::move(value))
+            , m_flag(flag)
+        {
+        }
+
+        constexpr optional(std::add_lvalue_reference_t<T> value, flag_type&& flag)
+            : m_value(value)
+            , m_flag(std::move(flag))
+        {
+        }
+
+        constexpr self_type& operator=(std::nullopt_t)
+        {
+            m_flag = false;
+            return *this;
+        }
+
+        template <class TO>
+        requires(
+            not std::same_as<self_type, TO> and
+            std::assignable_from<std::add_lvalue_reference_t<T>, TO>
+        )
+        constexpr self_type& operator=(TO&& rhs)
+        {
+            m_value = std::forward<TO>(rhs);
+            m_flag = true;
+            return *this;
+        }
+
+        constexpr self_type& operator=(const self_type& rhs)
+        {
+            m_value = rhs.m_value;
+            m_flag = rhs.m_flag;
+            return *this;
+        }
+
+        template <class TO, boolean_like BO>
+        requires(
+            util::both_assignable_from_cref<T, TO, B, BO> and
+            not util::initializable_from_refs<T, optional<TO, BO>> and
+            not util::assignable_from_refs<T, optional<TO, BO>>
+        )
+        constexpr self_type& operator=(const optional<TO, BO>& rhs)
+        {
+            m_value = rhs.m_value;
+            m_flag = rhs.m_flag;
+            return *this;
+        }
+
+        constexpr self_type& operator=(self_type&& rhs)
+        {
+            m_value = std::move(rhs.m_value);
+            m_flag = std::move(rhs.m_flag);
+            return *this;
+        }
+
+        template <class TO, boolean_like BO>
+        requires(
+            util::both_assignable_from_cond_ref<T, TO, B, BO> and
+            not util::initializable_from_refs<T, optional<TO, BO>> and
+            not util::assignable_from_refs<T, optional<TO, BO>>
+        )
+        constexpr self_type& operator=(optional<TO, BO>&& rhs)
+        {
+            m_value = std::move(rhs.m_value);
+            m_flag = std::move(rhs.m_flag);
+            return *this;
+        }
+
+        constexpr bool has_value() const noexcept;
```

Comment:
I prefer to keep the declaration and implementation separated, so that it's easier to read the whole interface of a class. However, I wanted to avoid repeating the constraints of the different constructors, it would have been cumbersome.